cpg_workflows/targets.py

```diff
@@ -97,18 +97,118 @@ def rich_id_map(self) -> dict[str, str]:
+
+    def get_cohorts(self, only_active: bool = True) -> list['Cohort']:
+        """
+        Gets list of all cohorts.
+        Include only "active" cohorts (unless only_active is False)
+        """
+        cohorts = [cohort for _name, cohort in self._cohorts_by_name.items()]
+        if only_active:
+            cohorts = [c for c in cohorts if c.active and c.get_datasets()]
+        return cohorts
+
+    def get_cohort_by_name(self, name: str, only_active: bool = True) -> Optional['Cohort']:
+        """
+        Get cohort by name.
+        Include only "active" cohorts (unless only_active is False)
+        """
+        cohort = self._cohorts_by_name.get(name)
+        if not cohort:
+            logging.warning(f'Cohort {name} not found in the multi-cohort')
+            return None
+        if not only_active:  # Return cohort even if it's inactive
+            return cohort
+        if cohort.active and cohort.get_datasets():
+            return cohort
+        return None
+
+    def get_datasets(self, only_active: bool = True) -> list['Dataset']:
+        """
+        Gets list of all datasets.
+        Include only "active" datasets (unless only_active is False)
+        """
+        datasets = []
+        for cohort in self.get_cohorts(only_active):
+            datasets.extend(cohort.get_datasets(only_active))
+        return datasets
+
+    def get_sequencing_groups(self, only_active: bool = True) -> list['SequencingGroup']:
+        """
+        Gets a flat list of all sequencing groups from all cohorts.
+        Include only "active" sequencing groups (unless only_active is False)
+        """
+        all_sequencing_groups = []
+        for cohort in self.get_cohorts(only_active):
+            all_sequencing_groups.extend(cohort.get_sequencing_groups(only_active))
+        return all_sequencing_groups
+
+    def create_cohort(
+        self,
+        name: str,
+    ):
+        """
+        Create a cohort and add it to the multi-cohort.
+        """
+        if name in self._cohorts_by_name:
+            logging.debug(f'Cohort {name} already exists in the multi-cohort')
+            return self._cohorts_by_name[name]
+
+        c = Cohort(name=name, multicohort=self)
+        self._cohorts_by_name[c.name] = c
+        return c
+
+    def get_job_attrs(self) -> dict:
+        """
+        Attributes for Hail Batch job.
+        """
+        return {
+            'sequencing_groups': self.get_sequencing_group_ids(),
+            'datasets': [d.name for d in self.get_datasets()],
+            'cohorts': [c.name for c in self.get_cohorts()],
+        }
+
+
 class Cohort(Target):
     """
     Represents a "cohort" target - all sequencing groups from all datasets in the workflow.
     Analysis dataset name is required and will be used as the default name for the
     cohort.
     """
 
-    def __init__(self) -> None:
+    def __init__(self, name: str | None = None, multicohort: MultiCohort | None = None) -> None:
```

Comment:
Why does a cohort take a multicohort as an input? Is this to do with getting cohort Stage input from a multicohort stage? I'm assuming (still reading) that we go to metamist, get all cohorts in the analysis (MultiCohort), for each of those create a Cohort, then clustering by Dataset we create the Dataset input types...

Reply:
Every target takes in its 'parent' as input in the constructor so that you can link them. 
You can see this behaviour in SG's taking in a Dataset and Datasets taking in a Cohort. 

> Is this to do with getting cohort Stage input from a multicohort stage? 
More so used the otherway around at the moment. 

Sorry I'm not sure I understand the second part of your message.

Reply:
Makes sense to me! The other part is probably just incoherent rambling